design/proof-erasure-coding.md

```diff
@@ -0,0 +1,419 @@
+     |  |  |  ^  |  |  |  |     |  |  |  |  |  |  |  |
+     |  |  |  |  |  |  |  |     |  |  |  |  |  |  |  |
+     |  |  |  |  |  |  |  |     v  v  v  v  v  v  v  v
+
+    -------------------------  -------------------------
+    |//|//|//|//|//|//|//|//|  |//|//|//|//|//|//|//|//|
+    -------------------------  -------------------------
+
+         original data                  parity
+
+
+This only works for small amounts of data. When the polynomial is for instance
+defined over byte sized elements from a Galois field of 2^8, you can only encode
+2^8 = 256 bytes (data and parity combined).
+
+Interleaving
+------------
+
+To encode larger pieces of data with erasure coding, interleaving is used. This
+works by taking larger shards of data, and encoding smaller elements from these
+shards.
+
+    data shards
+
+    -------------    -------------    -------------    -------------
+    |x| | | | | |    |x| | | | | |    |x| | | | | |    |x| | | | | |
+    -------------    -------------    -------------    -------------
+     |                /                /                |
+      \___________   |   _____________/                 |
+                  \  |  /  ____________________________/
+                   | | |  /
+                   v v v v
+
+                  ---------         ---------
+            data  |x|x|x|x|   -->   |p|p|p|p|  parity
+                  ---------         ---------
+
+                                     | | | |
+       _____________________________/ /  |  \_________
+      /                 _____________/   |             \
+     |                 /                /               |
+     v                v                v                v
+    -------------    -------------    -------------    -------------
+    |p| | | | | |    |p| | | | | |    |p| | | | | |    |p| | | | | |
+    -------------    -------------    -------------    -------------
+
+    parity shards
+
+This is repeated for each element inside the shards. In this manner, we can
+employ erasure coding on a Galois field of 2^8 to encode 256 shards of data, no
+matter how big the shards are.
+
+The amount of original data shards is typically called K, the amount of parity
+shards M, and the total amount of shards N.
+
+Adversarial erasure
+-------------------
+
+The disadvantage of interleaving is that it weakens the protection against
+adversarial erasure that Reed-Solomon provides.
+
+An adversary can now strategically remove only the first element from more than
+half of the shards, and the dataset will be damaged beyond repair. For example,
+with a dataset of 1TB erasure coded into 256 data and parity shards, an
+adversary could strategically remove 129 bytes, and the data can no longer be
+fully recovered.
+
+Implications for storage proofs
+-------------------------------
+
+This means that when we check for missing data, we should perform our checks on
+entire shards to protect against adversarial erasure. In the case of our Merkle
```

Comment:
I didn't want to deviate from the reasoning in the "Compact Proofs of Retrievabilty" paper. This gives us probabilistic certainty that the slot data is still there. We might be able to loosen the proof requirements because they are covered by incentives, but that becomes much harder to reason about, I think.